src/main/java/org/cryptomator/cryptofs/health/shortened/ShortenedNamesCheck.java

```diff
@@ -0,0 +1,131 @@
+import java.nio.file.StandardOpenOption;
+import java.nio.file.attribute.BasicFileAttributeView;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.util.Set;
+import java.util.function.Consumer;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+import static org.cryptomator.cryptofs.common.Constants.DEFLATED_FILE_SUFFIX;
+import static org.cryptomator.cryptofs.common.Constants.INFLATED_FILE_NAME;
+
+/**
+ * Visits all c9s directories and checks if they all are valid shortened resource according the Cryptomator vault specification.
+ */
+public class ShortenedNamesCheck implements HealthCheck {
+
+	private static final Logger LOG = LoggerFactory.getLogger(ShortenedNamesCheck.class);
+	private static final int MAX_TRAVERSAL_DEPTH = 3;
+	private static final BaseEncoding BASE64URL = BaseEncoding.base64Url();
+
+	@Override
+	public String name() {
+		return "Shortened Names Check";
+	}
+
+	@Override
+	public void check(Path pathToVault, VaultConfig config, Masterkey masterkey, Cryptor cryptor, Consumer<DiagnosticResult> resultCollector) {
+
+		// scan vault structure:
+		var dataDirPath = pathToVault.resolve(Constants.DATA_DIR_NAME);
+		var dirVisitor = new ShortenedNamesCheck.DirVisitor(resultCollector);
+		try {
+			Files.walkFileTree(dataDirPath, Set.of(), MAX_TRAVERSAL_DEPTH, dirVisitor);
+		} catch (IOException e) {
+			LOG.error("Traversal of data dir failed.", e);
+			resultCollector.accept(new CheckFailed("Traversal of data dir failed. See log for details."));
+		}
+	}
+
+	// visible for testing
+	static class DirVisitor extends SimpleFileVisitor<Path> {
+
+		private final Consumer<DiagnosticResult> resultCollector;
+
+		public DirVisitor(Consumer<DiagnosticResult> resultCollector) {
+			this.resultCollector = resultCollector;
+		}
+
+		@Override
+		public FileVisitResult visitFile(Path dir, BasicFileAttributes attrs) throws IOException {
+			var name = dir.getFileName().toString();
+			if (attrs.isDirectory() && name.endsWith(Constants.DEFLATED_FILE_SUFFIX)) {
+				checkShortenedName(dir);
+			}
+			return FileVisitResult.CONTINUE;
+		}
+
+		// visible for testing
+		//TODO: check if content of longName is decryptable. If not, result is Missing (not Mismatch as right now)
+		// dirID is needed for that
+		void checkShortenedName(Path dir) throws IOException {
+			Path nameFile = dir.resolve(INFLATED_FILE_NAME);
+
+			final BasicFileAttributes attrs;
+			try {
+				attrs = Files.getFileAttributeView(nameFile, BasicFileAttributeView.class, LinkOption.NOFOLLOW_LINKS).readAttributes();
+			} catch (NoSuchFileException e) {
+				resultCollector.accept(new MissingLongName(dir));
+				return;
+			}
+			if (!attrs.isRegularFile()) {
+				resultCollector.accept(new MissingLongName(dir));
+				return;
+			} else if (attrs.size() > LongFileNameProvider.MAX_FILENAME_BUFFER_SIZE) {
+				resultCollector.accept(new ObeseNameFile(nameFile, attrs.size()));
+				return;
+			}
+
+			var longName = readLongName(nameFile);
```

Comment:
Since we can now facilitate new APIs, let's get rid of `readLongName (...)` \*

```suggestion
			var longName = Files.readString(nameFile, UTF_8);
```

\* just here, not in LongFileNameProvider - that can be done as a separate commit directly on develop eventually

Reply:
Just a remark about the future replacement in LongFileNameProvider:  `Files.readString(...)` has no size restrictions of the file to read. This should be kept in mind. Here it does not matter, because the size is checked beforehand.